### PR TITLE
Automate server removal from NewRelic and ELB and server shutdown

### PR DESCRIPTION
**Pull request by philbogle**
This utility class eliminates most of gruntwork to deprovision servers, allowing you to bulk disable NewRelic monitoring, unregister from the load balancer, and terminate instances. At the same time it includes a number of safeguards to make mistake less likely.

A sample invocation looks like this; it takes a glob describing the frontends to update
`ServerTools.deprovision_frontends('frontend-hoc??')`

The tool displays a confirmation prompt listing the servers that will be affected:

```
Are you sure you want to deregister, disable alerts, and terminate the EC2 instances for 
frontend-hoc10, frontend-hoc19, frontend-hoc13, frontend-hoc20, ...
```

`deprovision_frontends` refuses to run if of the matching frontends are still listed iany frontends list in the app-servers attribute of cdo-secrets;  this ensures that servers aren't accidentally left behind in the chef config and provides a double-check that the user is removing the servers they expect to.

It also refuses to run if the wildcard matches all of the remaining servers.

There are also finer grained commands for removing wildcarded servers from the load balancer, from NewRelic monitoring, and for terminating the instances.  The interactive versions of these commands likewiseise display prompt.

I removed the automatic termination protection in the script for creating new instances.  However, we will retain that protection for a core of existing instances so we don't have to worry about terminating all of our instances via the tool even if all else fails.

Using the tool requires a little extra setup (installing the aws tools) which I think is fine because we want it to be used by power users only.

Sample irb session:

```
ServerTools.deprovision_frontends('frontend-hoc??')

Are you sure you want to deregister, disable alerts, and terminate the EC2 instances for frontend-hoc10, frontend-hoc19, frontend-hoc13, frontend-hoc20, frontend-hoc21, frontend-hoc15, frontend-hoc12, frontend-hoc14, frontend-hoc22, frontend-hoc24, frontend-hoc16, frontend-hoc18, frontend-hoc11, frontend-hoc23? (y/n) y

Deregistering i-bc5ba50c, i-513fddef, i-c3c1723a, i-ca1d9e7d, i-78a31081, i-489e7cf6, i-e44caf5a, i-98897028, i-3c71888c, i-40a123f7, i-ffb33048, i-d9d72e69, i-710a8ac6, i-67b558d9 on production-dashboard, production-pegasus, production-redirects

Disabling NewRelic alerts for frontend-hoc10 frontend-hoc19 frontend-hoc13 frontend-hoc20 frontend-hoc21 frontend-hoc15 frontend-hoc12 frontend-hoc14 frontend-hoc22 frontend-hoc24 frontend-hoc16 frontend-hoc18 frontend-hoc11 frontend-hoc23

Terminating frontend-hoc10 frontend-hoc19 frontend-hoc13 frontend-hoc20 frontend-hoc21 frontend-hoc15 frontend-hoc12 frontend-hoc14 frontend-hoc22 frontend-hoc24 frontend-hoc16 frontend-hoc18 frontend-hoc11 frontend-hoc23

Instance ids: i-bc5ba50c i-513fddef i-c3c1723a i-ca1d9e7d i-78a31081 i-489e7cf6 i-e44caf5a i-98897028 i-3c71888c i-40a123f7 i-ffb33048 i-d9d72e69 i-710a8ac6 i-67b558d9)
```

### CODE REVIEW

**Comment by philbogle** [comment]
@mehalshah  FYI

**Comment by philbogle** [comment]
I am making some improvements, will add you back when ready.

**Comment by philbogle** [comment]
@wjordan This is ready to look at now

File: bin/add_frontend_server.rb [hunk 1]
```diff
@@ -182,8 +182,8 @@ def generate_instance_zone_and_name(ec2client, ssh_username, frontend_name = nil
                                                     monitoring: {
                                                         enabled: true
                                                     },
-                                                    # Prevent api termination, except for adhoc instances.
-                                                    disable_api_termination: (environment != 'adhoc'),
+                                                    # Don't disable api termination.
+                                                    disable_api_termination: false,
```
Comment (mehalshah) [inline bin/add_frontend_server.rb:186]: Maybe make this user configurable?

Comment (philbogle) [inline bin/add_frontend_server.rb:186]: Yes, I will do that in my next PR.  (This one is pretty big and multifaceted already)

File: lib/cdo/newrelic.rb [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  # "Index" an array of hashes by return a map from the value of the given field
+  # to the corresponding record. Field should be unique to avoid collisions.
+  def index(array, field)
+    {}.tap do |result|
+      array.each do |item|
+        raise "#{field} is not unique" unless result[item[field]].nil?
+        result[item[field]] = item
+      end
+    end
+  end
```
Comment (wjordan) [inline lib/cdo/newrelic.rb:121]: `#group_by` could be used here:

```
array = [{id: 1, x: 'a'}, {id: 2, x: 'b'}]
> array.group_by{|x|x[:id]}
=> {1=>[{:id=>1, :x=>"a"}], 2=>[{:id=>2, :x=>"b"}]}
```

but to enforce uniqueness and collapse the value-array into a single object you'd need to do some extra steps:

```
raise "#{field} is not unique" unless array.uniq!{|x|x[field]}.nil?
array.group_by{|x|x[field]}.map{|k,v| [k,v.first]}.to_h
```

or using [`Hash#transform_values`](https://github.com/rails/rails/blob/4-2-stable/activesupport/lib/active_support/core_ext/hash/transform_values.rb) extension from ActiveSupport:

```
require 'active_support/core_ext/hash/transform_values'
raise "#{field} is not unique" unless array.uniq!{|x|x[field]}.nil?
array.group_by{|x|x[:id]}.transform_values{|v|v.first}
```

so not much more concise in the end

**Comment by wjordan** [comment]
LGTM, I think these scripts look quite helpful for our maintenance tasks.

One thing I was wondering about- would it be possible for us to start using the Chef-managed node list as our source of truth of active frontend app-servers, rather than going through the extra trouble of maintaining and synchronizing the extra `cdo-apps.app_servers` attribute that Chef's `knife` command knows nothing about? I think it could make a lot of our frontend-management a lot simpler to remove this extra layer of config.

File: lib/cdo/newrelic.rb (hunk 2 above)
Comment (philbogle) [inline lib/cdo/newrelic.rb:121]: I do like your version a bit better, but not enough to retest things given we're in a bit of time crunch before HOC.  (Unfortunately, it's challenging to add proper automated integration tests for this kind of Chef/NewRelic/AWS automation; I'm not sure the payoff would justify the effort.)

**Comment by philbogle** [comment]
I really like the idea of making the Chef-managed the source of truth regardling active frontend servers-- having this information in two places is not good at all.  

This is a good technical debt item to work on during HOC if all is quiet or in the following week.

**Comment by philbogle** [comment]
The CircleCI failure is spurious since this PR touches only standalone scripts and libraries not used in the any of the CircleCI tests.

**Pull request merged**